app/api.go

```diff
@@ -235,18 +236,26 @@ func (api *API) addHandler(r *http.Request) (interface{}, error) {
 
 	// See if this is a valid RekorEntry
 	rekorEntry, err := types.ParseRekorEntry(byteLeaf)
-	if err == nil {
-		b, err := rekorEntry.MarshalledLeaf()
-		if err != nil {
-			return nil, err
-		}
-		byteLeaf = b
-	} else {
-		logging.Logger.Infof("Not a valid rekor entry: %s", err)
+	if err != nil {
+		logging.Logger.Errorf("Not a valid rekor entry: %s", err)
+		return nil, err
 	}
-
+	// Check to see if the entry already exists
 	server := serverInstance(api.logClient, api.tLogID)
+	b, err := rekorEntry.MarshalledLeaf()
+	if err != nil {
+		return nil, err
+	}
+	if resp, err := server.getLeaf(b, api.tLogID); err != nil && len(resp.getLeafResult.Leaves) != 0 {
+		return addResponse{
+			Status: RespStatusCode{Code: getGprcCode(codes.AlreadyExists)},
+		}, nil
+	}
```

Comment:
Is this to save from uploading first and then checking if the tarball already exists? (save bandwidth / speed of operation?)

Reply:
Yeah exactly. The linux kernel validation can take several minutes.